metadata/src/main/java/org/apache/kafka/metadata/BrokerRegistration.java

```diff
@@ -213,12 +230,30 @@ public String toString() {
         bld.append("}");
         bld.append(", rack=").append(rack);
         bld.append(", fenced=").append(fenced);
+        bld.append(", inControlledShutdown=").append(inControlledShutdown);
         bld.append(")");
         return bld.toString();
     }
 
-    public BrokerRegistration cloneWithFencing(boolean fencing) {
-        return new BrokerRegistration(id, epoch, incarnationId, listeners,
-            supportedFeatures, rack, fencing);
+    public Optional<BrokerRegistration> maybeCloneWith(
+        Optional<Boolean> fencingChange,
+        Optional<Boolean> inControlledShutdownChange
+    ) {
+        boolean newFenced = fencingChange.orElse(fenced);
+        boolean newInControlledShutdownChange = inControlledShutdownChange.orElse(inControlledShutdown);
+
+        if (newFenced == fenced && newInControlledShutdownChange == inControlledShutdown)
+            return Optional.empty();
```

Comment:
Okay but we could just return `this`. Outside of `supportedFeatures`, `BrokerRegistration` is immutable and it override `equals` and `hashCode` so this should be safe.

I'll submit a minor PR to make `supportFeature` immutable.

Reply:
Yeah, you're right. I am not sure why I use this Optional here.